### ISSUE TITLE
PyISY should process all status changes immediately after reception

### ISSUE THREAD

**Issue by altersis** [automicus/PyISY#51]
Automations that depend on ISY events sometimes take up to 25 seconds to trigger. This is specially evident for devices that emit one single event under normal circumstances. A good example is the Remote Linc 2 keypad; when configured to emit a DON-only on key presses, the ISY will send only a DON event. After only  5 or 6 seconds users start to think that the keypress is lost and will press on again. This will trigger the DON event quickly, but a second DON event will trigger  after a maximum time of 25 seconds.

**Comment by altersis** [comment]
In-depth analysis of the code used to process incoming events from the ISY shows that the events.watch() assumes every data element inside an update from the ISY is delimited with a newline character '\n', which is incorrect: only the HTTP header is. Once a subscription is initialized, the payload of each update from the ISY to PyISY contains an XML document that is delimited by the <?xml> and the <Event> tags. The subscription confirmation message contains data that is not relevant, and for that reason can be ignored. 

In addition, Reads from the input buffer are implemented in events.read(), which uses a file object to perform a readline(), which implies extracting data in small increments.

Solution:

Modify watch() to read the buffer in 4 KB blocks, skipping the subscription confirmation message if present, then incrementally extract each update message by the proper XML tags, until the buffer is completely processed.

**Comment by shbatm** [comment]
Just an observation (reading sockets/streams is not my forte), it looks like EventsSSL already implements a 4 kb buffer but Events does not.

**Comment by altersis** [comment]
Hey @shbatm, thankis for chiming in. This is one of those problems that are super tricky to find, I checked the EventsSSL, and although I cannot test it (I do not use SSL in my ISY) we'll most likely need to change the parsing to be done by XML tags as well. I exchanged a couple of messages with @OverloadUT and we'll do our best to chat tomorrow so I can understand the work already done to ensure that whatever I suggest to change jives with your plans. you can check the non-ssl change in my fork of automicus' PyISY.

**Comment by rmkraus** [comment]
Some historical context to this one.

Early versions of this library only contained Events. Some time later, I wrote EventsSSL using lessons learned from writing Events. As a result, EventsSSL is MUCH better, has better performance all around, and does fewer dumb things (like extending the socket base class because why is that necessary). I always wanted to rewrite Events to use the same code as EventsSSL and stop maintaining two different classes. Then I didn't.

@altersis As a workaround, try enabling SSL on your ISY. See if this fixes the issues you are having. If it does, then the easiest way to fix this would be to replace Events with the existing EventsSSL and strip out the encryption bits. Basically, just don't wrap the socket with the TLS context and Bob's your uncle (probably).

Hopefully this helps.

**Comment by shbatm** [comment]
@altersis: I agree with @overloadut's plan of getting your changes implemented ASAP. After that I'll work on getting them merged into my changes. 

I did combine the two events classes in my PR (including removing the subclassing @rmkraus mentioned) but the one thing I didn't change was the difference in how ssl and non-ssl connections read from the socket. Mostly because I don't have a great understanding of sockets and when I tried using the SSL method directly with non-SSL, it broke my updates.

Looking forward to looking through your update and again, great catch!

**Comment by altersis** [comment]
@rmkraus, @shbatm, @OverloadUT: thanks for getting involved. I would like to thank you for the work you've done on this, it's phenomenal! I tried enabling SSL on my ISY994, but count not make HASS connect using HTTPS:, it always pings the ISY with HTTPS, and then falls back to HTTP. I never got a certificate for the ISY, so I suspect this is the reason HASS cannot connect via HTTPS:

`2019-07-19 10:53:11 INFO (SyncWorker_13) [homeassistant.components.isy994] ISY Request: https://192.168.1.16:80/rest/ping
2019-07-19 10:53:21 WARNING (MainThread) [homeassistant.setup] Setup of isy994 is taking over 10 seconds.
2019-07-19 10:53:21 ERROR (SyncWorker_13) [homeassistant.components.isy994] Timed out waiting for response from the ISY device.
2019-07-19 10:53:21 INFO (SyncWorker_13) [homeassistant.components.isy994] ISY Request: http://192.168.1.16:80/rest/ping
2019-07-19 10:53:21 INFO (SyncWorker_13) [homeassistant.components.isy994] ISY Response Recieved
2019-07-19 10:53:21 WARNING (SyncWorker_13) [homeassistant.components.isy994] PyISY could not connect with the controller. Trying again with HTTP.
2019-07-19 10:53:21 INFO (SyncWorker_13) [homeassistant.components.isy994] ISY Request: http://192.168.1.16:80/rest/config
2019-07-19 10:53:21 INFO (SyncWorker_13) [homeassistant.components.isy994] ISY Response Received`

I did look at EventSSL.watch(), and it looks like it should work, although I would like to understand: while checking, I noticed that the code for read() is:

> 

    def read(self):
        if self._reader is None:
            self._NIYerror()
        else:
            loop = True
            output = ''
            while loop:
                try:
                    new_data = self.socket.recv(4096)
                except ssl.SSLWantReadError:
                    pass
                except socket.error:
                    loop = False
                else:
                    if sys.version_info.major == 3:
                        new_data = new_data.decode('utf-8')
                    output += new_data
                    if len(new_data) * 8 < 4096:
                        loop = False

            return output.split('\n')

> 

1. what's the reason for doing the decode when running python 3?
2. what's the thinking behind the exit condition 

>  if len(new_data) * 8 < 4096:
                        loop = False

**Comment by rmkraus** [comment]
1. I don't remember the exact reason, but it almost definitely had to do with the ISY passing in the degrees symbol when using the Weather module.
2. There are two exit conditions. 1) There is an unexpected socket error. This typically will mean the connection is gone and there is nothing left for us to read. 2) I asked for 4k of data and you gave me less than that. This means I have reached the end of the socket buffer and the end of the message.

**Comment by jimboca** [comment]
While this work is being done are there any changes needed to support the chunked encoding method for ISY v5, which will become the default someday?

**Comment by shbatm** [comment]
@altersis re: the '`.decode()`' function, IIRC I encountered this somewhere else upgrading an app from Python 2 to 3--it has something to do with Python 3 not automatically converting bytes to strings.

**Comment by shbatm** [comment]
@jimboca -- I'll take a stab at including this in #44 so we don't hold up a fix for this.

**Comment by altersis** [comment]
Thanks @rmkraus, I'll dig a bit more into the encoding based on @shbatm comment.  The only question that I still have is: why is it necessary to multiply the length of the new data by 8 to compare it with the 4096 bytes?

**Comment by rmkraus** [comment]
I'm sorry, I missed that part of your question.

len(string) -> number of characters per string
8 -> number of bits per character
len(string) * 8 -> number of bits in the string
Then you compare that to 4096 bits

**Comment by altersis** [comment]
Hmmm... I might be missing something here, but we just need to compare the number of characters in the string against the maximum number of bytes (4096 bytes or 4KB) in the buffer, no? that would be:
```
if len(new_data) < 4096:
  loop = False
```
makes sense?

**Comment by OverloadUT** [comment]
I share your confusion. Here's the chain of logic:
`socket.recv(4096)` -> This should give us 4 kilobytes (not bits)
`decode('utf-8')` -> From my understanding, this should give us a string with one character for every *4 bytes*. So those 4096 bytes have now turned in to 1024 characters
`len(new_data)` -> This should give us 1024 now, right?

Now that I've typed that out, shouldn't we be doing `if len(new_data) * 4 < 4096:` because UTF-8 essentially divided by 4, not 8? Such low-level encoding is not my wheelhouse at all, so I'm grasping at straws to understand this.


This actually raises something that I have always wondered about this part of the PyISY library: should we really be doing all of this super low level `socket` stuff ourselves? It feels to me that we ought to be using a stable well-established websocket library for Python so that we don't have to handle such low-level concerns.

Of course that would be a larger project and perhaps there are good reasons for doing it manually...

**Comment by shbatm** [comment]
@OverloadUT yep now I'm confused too.

Looking at [this example](http://danielhnyk.cz/simple-server-client-aplication-python-3/) we can test the raw received data before decoding:

```python
    def read(self):
        """Read data from the socket."""
        BUFFER_SIZE = 4096
        loop = True
        output = ''
        while loop:
            try:
                new_data = self.socket.recv(BUFFER_SIZE)
            except ssl.SSLWantReadError:
                pass
            except socket.error:
                loop = False
            else:
                siz = sys.getsizeof(new_data)
                if sys.version_info.major == 3:
                    new_data = new_data.decode('utf-8')
                output += new_data
                if siz < BUFFER_SIZE:
                    loop = False

        return output.split('\n')
```

**Comment by rmkraus** [comment]
You cannot use sys.getsizeof. That returns the size of the data + all of Python's extra variable information. 

Looks like you might be right. Although the Python docs do not specify, according to folklore, it appears that socket.recv does specify kilobytes. Easiest way to test would be to drop a pdb.settrace() in right after the socket.recv and run a len on new_data. Is it 4096 or is it 512?

I don't have an ISY hooked up so I unfortunately cannot help here. 

The reason for not using an external library was because I didn't feel it was necessary at the time. If we were managing many websockets that randomly opened and closed, that would be one thing. For a single long-lived websocket, adding a dependency seems silly.

**Comment by shbatm** [comment]
OK RE: getsizeof.  I'll revert in my PR and use a comparison.  

@OverloadUT -- Just FYI, found this tidbit when I was Googling around:

> For unicode strings the per-character size goes up to 2 or 4 (depending on compilation options). On Python 3.3 and newer, Unicode strings take up between 1 and 4 bytes per character, depending on the contents of the string.

**Comment by shbatm** [comment]
Updated #52 to use `len` instead of `getsizeof` and reordered per post above to check size before any decoding is done.

@rmkraus was right, you get different values for `len` and `getsizeof` and `new_data` is pulled in as 4096 bytes:

```shell
2019-07-19 15:31:09 WARNING (Thread-3) [custom_components.isy994] Recieved 271 bytes, getsizeof reported 304.
2019-07-19 15:31:09 WARNING (Thread-3) [custom_components.isy994] Recieved 215 bytes, getsizeof reported 248.
2019-07-19 15:31:09 WARNING (Thread-3) [custom_components.isy994] Recieved 260 bytes, getsizeof reported 293.
2019-07-19 15:31:09 WARNING (Thread-3) [custom_components.isy994] Recieved 712 bytes, getsizeof reported 745.
2019-07-19 15:31:09 WARNING (Thread-3) [custom_components.isy994] Recieved 255 bytes, getsizeof reported 288.
2019-07-19 15:31:10 WARNING (Thread-3) [custom_components.isy994] Recieved 4096 bytes, getsizeof reported 4129.
2019-07-19 15:31:10 WARNING (Thread-3) [custom_components.isy994] Recieved 4096 bytes, getsizeof reported 4129.
2019-07-19 15:31:10 WARNING (Thread-3) [custom_components.isy994] Recieved 4096 bytes, getsizeof reported 4129.
2019-07-19 15:31:10 WARNING (Thread-3) [custom_components.isy994] Recieved 4096 bytes, getsizeof reported 4129.
2019-07-19 15:31:10 WARNING (Thread-3) [custom_components.isy994] Recieved 4096 bytes, getsizeof reported 4129.
2019-07-19 15:31:10 WARNING (Thread-3) [custom_components.isy994] Recieved 4096 bytes, getsizeof reported 4129.
2019-07-19 15:31:10 WARNING (Thread-3) [custom_components.isy994] Recieved 4096 bytes, getsizeof reported 4129.
2019-07-19 15:31:10 WARNING (Thread-3) [custom_components.isy994] Recieved 4096 bytes, getsizeof reported 4129.
2019-07-19 15:31:10 WARNING (Thread-3) [custom_components.isy994] Recieved 4096 bytes, getsizeof reported 4129.
2019-07-19 15:31:10 WARNING (Thread-3) [custom_components.isy994] Recieved 4096 bytes, getsizeof reported 4129.
2019-07-19 15:31:10 WARNING (Thread-3) [custom_components.isy994] Recieved 4096 bytes, getsizeof reported 4129.
2019-07-19 15:31:10 WARNING (Thread-3) [custom_components.isy994] Recieved 4096 bytes, getsizeof reported 4129.
2019-07-19 15:31:10 WARNING (Thread-3) [custom_components.isy994] Recieved 4096 bytes, getsizeof reported 4129.
2019-07-19 15:31:10 WARNING (Thread-3) [custom_components.isy994] Recieved 4096 bytes, getsizeof reported 4129.
2019-07-19 15:31:10 WARNING (Thread-3) [custom_components.isy994] Recieved 4096 bytes, getsizeof reported 4129.
2019-07-19 15:31:10 WARNING (Thread-3) [custom_components.isy994] Recieved 4096 bytes, getsizeof reported 4129.
2019-07-19 15:31:10 WARNING (Thread-3) [custom_components.isy994] Recieved 4096 bytes, getsizeof reported 4129.
2019-07-19 15:31:10 WARNING (Thread-3) [custom_components.isy994] Recieved 4096 bytes, getsizeof reported 4129.
2019-07-19 15:31:10 WARNING (Thread-3) [custom_components.isy994] Recieved 4096 bytes, getsizeof reported 4129.
2019-07-19 15:31:10 WARNING (Thread-3) [custom_components.isy994] Recieved 4096 bytes, getsizeof reported 4129.
2019-07-19 15:31:10 WARNING (Thread-3) [custom_components.isy994] Recieved 4096 bytes, getsizeof reported 4129.
2019-07-19 15:31:10 WARNING (Thread-3) [custom_components.isy994] Recieved 4096 bytes, getsizeof reported 4129.
2019-07-19 15:31:10 WARNING (Thread-3) [custom_components.isy994] Recieved 4096 bytes, getsizeof reported 4129.
2019-07-19 15:31:10 WARNING (Thread-3) [custom_components.isy994] Recieved 4096 bytes, getsizeof reported 4129.
2019-07-19 15:31:10 WARNING (Thread-3) [custom_components.isy994] Recieved 4096 bytes, getsizeof reported 4129.
2019-07-19 15:31:10 WARNING (Thread-3) [custom_components.isy994] Recieved 1477 bytes, getsizeof reported 1510.
2019-07-19 15:31:11 WARNING (Thread-3) [custom_components.isy994] Recieved 319 bytes, getsizeof reported 352.
```

**Comment by altersis** [comment]
Gents, can I suggest a simpler approach?

the idea would be use the following in watch() immediately after the check for data:

                if self in inready:
                   try:
                     self.parent.log.debug('PyISY: data in buffer')
                     tmp_buff =  self._reader.read()
                   except socket.error:
                     self.parent.log.error('PyISY: could not read from event stream.')
                     continue
                   currPos  = tmp_buff.find('</s:Envelope>') + 13
                   if currPos == -1:
                     currPos = 0
                   while currPos < len(tmp_buff):
                     eventStart = tmp_buff.find('<?xml', currPos)
                     eventEnd   = tmp_buff.find('</Event>', currPos) + 8
                     if eventStart > -1 and eventEnd > eventStart:
                       data = tmp_buff[eventStart : eventEnd]
                       self.parent.log.debug('PyISY: routing data')
                       self._routemsg(data)
                       currPos = eventEnd + 1
                     else:
                       self.parent.log.warning('PyISY: Malformed event data: '+tmp_buff[currPos:])
                       currPos = len(tmp_buff)
                   self.parent.log.debug('PyISY: Buffer processed\n')

in this way, we do not have to worry about buffers and sizes, no need for read(), less resources to execute... abd adding the stuff for https: is super simple... what do you think?

**Comment by shbatm** [comment]
The concern I had about that was still using the stream reader--wouldn't it just continue to grow? (I'm not that familiar with Python's cleanup methods, but wouldn't this just continue to consume more memory the longer it ran? I could be way off base here.)

**Comment by altersis** [comment]
Right, we want to avoid the file reader! We can change one line to tmp_buff = self.socket.recv(BUFFER_SIZE) where BUFF_SIZE would be 4096, and were good... I'll try it in a moment. I've had the above code with the reader working for 4 days now, no problems... I have around 100 insteon devices (including around 15 motion sensors)  and they generate a good amount of insteon events per minute.

**Comment by altersis** [comment]
sorry, i meant: tmp_buff = str(self.recv(4096))

just tried it and it gives the same results. I'm going to let it run and monitor it

**Comment by shbatm** [comment]
If you did that, what happens when the message is bigger than 1 buffer (e.g. the initial event dump from the ISY), you wouldn't see `</Event>` in `tmp_buff` (or maybe you would only see the first one, I'd have to look at that initial dump again).

**Comment by altersis** [comment]
@shbatm, you have a very valid point, and also saw that need to get add more data to the present buffer if the current message is incomplete. when i first tried it with the reader, the buffers were big and there was no trouble during the initial data dump, but now that I checked the logs using the plain socket.recv(), this code runs faster, so the buffer is less full. I'll spend a few minutes to work on getting additional data if a message is incomplete. The regular operation once the data load is complete, has no problem

**Comment by altersis** [comment]
Hi guys, I think I have it working now, and the best is that I think this code already works for HTTPS and HTTP, although I have only tested it with HTTP. 

>                 if self in inready:
                   try:
                     self.parent.log.debug('PyISY: data in buffer')
                     tmp_buff = self.recv(4096).decode("utf-8")
                     #self.parent.log.debug('PyISY: buffer: \n'+tmp_buff)
                   except socket.error:
                     self.parent.log.error('PyISY: could not read from event stream.')
                     continue
                   currPos  = tmp_buff.find('</s:Envelope>') + 13
                   if currPos == -1:
                      currPos = 0
                   while currPos < len(tmp_buff):
                     eventStart = tmp_buff.find('<?xml', currPos)
                     eventEnd   = tmp_buff.find('</Event>', currPos) + 8
                     self.parent.log.debug('PyISY: start: '+str(eventStart)+' end: '+str(eventEnd)+" pos: "+str(currPos))
                     #self.parent.log.debug('PyISY: scope: '+tmp_buff[currPos:])
                     if eventStart > -1:       # Valid XML, process it
                        if eventEnd > eventStart:
                           data = tmp_buff[eventStart : eventEnd]
                           self.parent.log.debug('PyISY: routing data')
                           self._routemsg(data)
                           currPos = eventEnd
                        elif eventEnd == 7:   # Valid start of XML but no end, get the rest
                            tmp_buff = tmp_buff[eventStart:]
                            currPos = 0
                            self.parent.log.debug('PyISY: adding to buffer, incomplete message found')
                            try:
                              tmp_buff += self.recv(4096).decode("utf-8")
                            except socket.error:
                              self.parent.log.error('PyISY: could not read from event stream.')
                              break
                            if tmp_buff.find('</Event>') > -1:
                              continue
                            else:
                              self.parent.log.warning('PyISY: Malformed event data: '+tmp_buff[currPos:])
                              break
                        elif eventEnd < eventStart: # If we get here, we have BIG problems!
                            self.parent.log.error('PyISY: HELP, LOST MY POINTERS!')
                            break 
                     elif eventStart == -1 and eventEnd == 7: # no more XMLs found, exit
                          self.parent.log.debug('PyISY: no more events in buffer')
                          break
                   self.parent.log.debug('PyISY: Buffer processed\n')

The general idea is: Get up to a max of 4KB data, and within that buffer move a start and end of event references, processing a complete event before moving the references to the next one, until:

1. we hit the end of the messages, 
2. we only find the start of a new event, which means more data has to be brought from the socket. In this case we re-enter the loop and continue until we need more data or exit criteria 1 is met. 

I have tried it with my production environment here, and it gets the initial dump done super quick.

**Comment by OverloadUT** [comment]
Excellent! Get a PR up and I'll start testing and reviewing it on my end, or if you want to use #52 I'll just wait for shbatm to update :)

**Comment by rmkraus** [comment]
I really don't care for that. A socket reader should do just that, read sockets. I do not like the idea of building SOAP/XML logic that far down.

The logic should be:
1) wait for the socket to have data available (select statement)
2) Read data from the socket until there is no more data to be read (self.read)
3) parse the data from the socket (self._routemsg)

Any intermingling of that logic is bad for business. One of a many reasons for this is that the buffer should not be used as a medium term storage of data. As soon as data is in the socket buffer, it is the responsibility of the program connected to that socket to remove it. Letting data linger in the socket buffer simply because it is not part of the current message is a bad practice and bad hygiene. 

If there is a problem with partial messages being received, that should be handled in step 3 (this is not currently handled). If there is a problem with multiple messages being received at a time, that is a problem for step 3 (this is currently handled).

The problem identified by the issue was that events.py expected a newline after each message which delayed the processing of incoming messages. It looks as if #52 should resolve that issue. The only way I could think of to increase the efficiency of the current logic would be so run self._routemsg in a new thread so that self.watch can get back to reading the socket.  OR, if someone wanted to port this to asyncio. That would be the best option.

**Comment by altersis** [comment]
hi there, sorry I was away yesterday. I had some family commitments that I simply had to attend. We have @shbatm 's code already there, so why not get that merged to put out the fixes? It solves the primary problems and buys time to keep the discussion open to get a much better long-term solution... makes sense?

**Comment by rmkraus** [comment]
👍
